src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,377 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using System.Runtime.InteropServices;
+using Internal.IL.Stubs;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.IL
+{
+    public sealed class UnsafeAccessors
+    {
+        private const string InvalidUnsafeAccessorUsage = "Invalid usage of UnsafeAccessorAttribute.";
+
+        public static MethodIL TryGetIL(EcmaMethod method)
+        {
+            Debug.Assert(method != null);
+
+            CustomAttributeValue<TypeDesc>? decodedAttribute = method.GetDecodedCustomAttribute("System.Runtime.CompilerServices", "UnsafeAccessorAttribute");
+            if (!decodedAttribute.HasValue)
+            {
+                return null;
+            }
+
+            if (!TryParseUnsafeAccessorAttribute(method, decodedAttribute.Value, out UnsafeAccessorKind kind, out string name))
+            {
+                ThrowHelper.ThrowBadImageFormatException(InvalidUnsafeAccessorUsage);
+            }
+
+            GenerationContext context = new()
+            {
+                Kind = kind,
+                Declaration = method
+            };
+
+            MethodSignature sig = method.Signature;
+            TypeDesc retType = sig.ReturnType;
+            TypeDesc firstArgType = null;
+            if (sig.Length > 0)
+            {
+                firstArgType = sig[0];
+            }
+
+            // Using the kind type, perform the following:
+            //  1) Validate the basic type information from the signature.
+            //  2) Resolve the name to the appropriate member.
+            switch (kind)
+            {
+                case UnsafeAccessorKind.Constructor:
+                    // A return type is required for a constructor, otherwise
+                    // we don't know the type to construct.
+                    // The name is defined by the runtime and should be empty.
+                    if (retType.IsVoid || !string.IsNullOrEmpty(name))
+                    {
+                        ThrowHelper.ThrowBadImageFormatException(InvalidUnsafeAccessorUsage);
+                    }
+
+                    // Due to how some types degrade, we block on parameterized
+                    // types. For example ref or pointer.
+                    if ((retType.IsParameterizedType && !retType.IsArray)
+                        || retType.IsFunctionPointer)
+                    {
+                        ThrowHelper.ThrowBadImageFormatException(InvalidUnsafeAccessorUsage);
+                    }
+
+                    context.TargetType = ValidateTargetType(retType);
```

Comment:
```suggestion
                    // A return type is required for a constructor, otherwise
                    // we don't know the type to construct.
                    // The name is defined by the runtime and should be empty.
                    if (retType.IsVoid || retType.IsByRef || !string.IsNullOrEmpty(name))
                    {
                        ThrowHelper.ThrowBadImageFormatException(InvalidUnsafeAccessorUsage);
                    }

                    context.TargetType = ValidateTargetType(retType);
```

Reply:
No need to duplicate ValidateTargetType. It should be enough to just check for IsByRef.